RetroAchievements/RAWeb
# feat: set up Inertia i18n, use lang strings, attempt #2

This PR is a repeat of #2649, which was approved, but reverted in #2673 after it was found to being causing hydration issues related to server-side rendering.

I really like how easy the `laravel-react-i18n` library is, and I think our i18n/l10n solution needs to be so easy that we barely ever have to think about it.

The bug responsible for our hydration issue was reported in Dec 2023 here: https://github.com/EugeneMeles/laravel-react-i18n/issues/16. The bug is directly related to how the current major release of React (shipped in 2022) changed React's internal hydration algorithm to be more efficient.

I found the root cause of the hydration bug in the library itself and fixed it locally.

Upstream doesn't seem to be accepting PRs at the moment to fix the issue.

My original plan was to maintain a fork of the lib and add it to our _package.json_ file. However, that seems to cause other spurious errors in the RAWeb project due to peer dependency conflicts.

This PR implements my next-best idea, which is to maintain an internal fork of the lib with the patched changes. This is stored under _@/lib/laravel-react-i18n_. Most of this PR's diff is just pulling in the source code for the lib. Here's the summary of changes in this PR:

* Pull in a modified version of `laravel-react-i18n` into _@/lib/laravel-react-i18n_.
* Restore the changes from #2649.
* Pass down the user's locale setting in the server-side and client-side renders in _app.tsx_ and _ssr.tsx_.

Implementation details under _@/lib/laravel-react-i18n_ are safe to ignore.

---

**How to test this PR:**
* `sail npm install`
* `sail npm run build`
* `sail artisan inertia:start-ssr`
* Navigate to `/forums/recent-posts`.
* Observe that there is no hydration error or devtools console errors. Also navigate to `/game/1/hashes` and observe the same.
* In the database on your UserAccounts row, set your "locale" value to "pt_BR". Check `/forums/recent-posts` again.

en_US:
![Screenshot 2024-09-27 at 6 29 10 PM](https://github.com/user-attachments/assets/94ed34aa-6d7a-40ad-852a-7dec2dee2500)

pt_BR:
![Screenshot 2024-09-27 at 6 28 45 PM](https://github.com/user-attachments/assets/e1d61bfb-044b-4de3-a7d3-5b0a6833cb4e)

---

**Next steps:**
* Add a way for users to change their `locale` value via the settings page.
* Use the `t()` function on all React pages.
* After #2706 is merged, make sure the locale value is also flowing into our number and date formatting utils.
* Write some kind of documentation so native speakers can understand how to contribute translations.
